### PR TITLE
Fix/#1113 index file outside project directory

### LINKED ISSUE #1113: Index files outside project directory
Hi!

I was wondering if it is possible to index files outside project directory in order to have completion and code navigation. I'm working on projects that depend on other php files that are in a different location (these projects are not using composer).

Everyone here is using phpstorm and it is possible to add these files as 'External librairies' to have autocompletion of classes and methods and code navigation.

Thanks! :)

### PR DESCRIPTION
**Pull request by camilledejoye**
Fix https://github.com/phpactor/phpactor/issues/1113

Basically rename the constant and variables.
A little bonus is the human readable printing of the memory consumption in the console :smiley_cat: 

Should be merged in the same time as: https://github.com/phpactor/phpactor/pull/1123

### CODE REVIEW

File: lib/Extension/Command/IndexBuildCommand.php [hunk 1]
```diff
@@ -121,4 +121,14 @@ private function watch(OutputInterface $output): void
...
     }
+
+    /**
+     * @see https://www.php.net/manual/en/function.memory-get-usage.php#96280
+     */
+    private function formatMemory(int $memoryInBytes): string
+    {
+        $unit = ['B', 'KiB', 'MiB', 'GiB'];
+
+        return @round($memoryInBytes / pow(1024, ($i = floor(log($memoryInBytes, 1024)))), 2).' '.$unit[$i];
```
Comment (dantleech) [inline lib/Extension/Command/IndexBuildCommand.php]: could we use `sprintf('%s %s')` and remove the inline assign, why do we ignore the error from round?

Comment (dantleech) [reply lib/Extension/Command/IndexBuildCommand.php]: this "logic" would also need to be used by the language server indexer service.

does it add any value? we currently report this in megabytes - bytes and kilobytes don't matter - and gigabytes should be rare and would be represented as `1,000mb` which is fine

i think simplicity is fine for now.

File: lib/Extension/IndexerExtension.php [hunk 2]
```diff
@@ -57,7 +57,7 @@ class IndexerExtension implements Extension
     public const PARAM_INDEXER_BUFFER_TIME = 'indexer.buffer_time';
     public const PARAM_REFERENCES_DEEP_REFERENCES = 'indexer.reference_finder.deep';
     public const PARAM_IMPLEMENTATIONS_DEEP_REFERENCES = 'indexer.implementation_finder.deep';
-    public const PARAM_STUB_PATHS = 'indexer.stub_paths';
+    public const PARAM_EXTERNAL_SOURCE_PATHS = 'indexer.stub_paths';
```
Comment (dantleech) [inline lib/Extension/IndexerExtension.php:60]: We need to rename the string value too

**Comment by dantleech** [comment]
Thanks. I'm not sure about the `external_source_paths` - what about e.g. `external_roots` (we have `project_root`)?

We also need to rename all references to `stub paths` and ensure that the constant and the constant value match.

File: lib/Extension/Command/IndexBuildCommand.php (hunk 1 above)
Comment (dantleech) [reply lib/Extension/Command/IndexBuildCommand.php]: ok, i guess this is for the final memory usage report - i'd still be inclined to leave it as is.

Comment (camilledejoye) [reply lib/Extension/Command/IndexBuildCommand.php]: I simply copy/paste from the php docucmentation, link in the comment.
There is no value it's just prettier on the screen and easier to read.
For the `@` I don't know, maybe to protect against a division by zero or whatever. I don't really care about this like that, it's sugar it's not a code to maintain or anything so I admit I didn't look past the fact that it worked.
I will try to make it cleaner.

**Comment by camilledejoye** [comment]
> Thanks. I'm not sure about the external_source_paths - what about e.g. external_roots (we have project_root)?

I let you choose, I like the idea of external source because it match the exactly the use case I see for the feature.
But I guess we could consider that it will be used to reference an entire project which is included in a non traditional way.
The `root` part is also clearer to understand that everything under it will be indexed whereas with the `paths` we could wonder if it will be recursive or not.

> We also need to rename all references to stub paths and ensure that the constant and the constant value match.

I think I didn't miss any reference to the constant, only the value as you mentioned earlier.

**Comment by dantleech** [comment]
Some background: I took some time implementing this feature was because "stubs" can (and should) be shared between projects - we don't need the ~28mb and 5 second indexing overhead for each project - but finally it was faster to just accept the overhead and defer more efficient stub handling until later.

> But I guess we could consider that it will be used to reference an entire project

Actually there is an important difference here - the "stubs" are not watched. They are only indexed once.

Which raises the possibility of having two things: "static" sources (e.g. stubs, PEAR packages / whatever) and external paths which are somehow part of a project - or even having multiple project "roots" (so `indexer.project_root` becomes an array).

So if `stubs` refers to these "static" external sources, is there a better name then? Or would it be better to keep `stub_paths` and think about refactoring to allow multiple project roots (which will be watched) later?

So all in all I'm not very confident in any particular direction, I think `external_roots` and `project_root(s)` might make sense.

> I think I didn't miss any reference to the constant, only the value as you mentioned earlier.

The method names should also be updated.

File: lib/Extension/Command/IndexBuildCommand.php (hunk 1 above)
Comment (dantleech) [reply lib/Extension/Command/IndexBuildCommand.php]: well - it's code to maintain if it fails for whatever reason (copy / paste from PHP.net has got me in trouble before now)

**Comment by camilledejoye** [comment]
> Actually there is an important difference here - the "stubs" are not watched. They are only indexed once.

I didn't realized that... I guess the best option would be to do both:
1. Add the possibility to add new "roots" to be able to watch external sources
2. Add the possibility to add new "stubs" to be able to optimize the index time by indexing only once

This way you don't have to choose any specific direction and the user should be able to decide for himself.

For the `1.`, I have a preference for `external_roots`. I don't like the idea to mix the new feature with the existing `project_root`, it would be confusing to have it in plural in some places and singular in others.
And I think it's important to keep a strong separation between the current project the user is working on and the external sources it needs inside it.

File: lib/Extension/Command/IndexBuildCommand.php (hunk 1 above)
Comment (camilledejoye) [reply lib/Extension/Command/IndexBuildCommand.php]: > ok, i guess this is for the final memory usage report - i'd still be inclined to leave it as is.

So I revert the related commit ?

**Comment by dantleech** [comment]
Multiple roots corresponds to [workspace folders](https://microsoft.github.io/language-server-protocol/specification#workspace_workspaceFolders) in the language server protocol.

There are two "tasks" that I can see here:

1. Add new roots that are _watched_ is probably not too tricky.
2. Share stubs: 
  - Stubs would need to be indexed separately (first on not existing, then periodically or manually)
  - We might need to "merge" indexes (at runtime): which is really not trivial with the current architecture.  

(1) is probably not to difficult (but it could be a can of worms) - 
(2) I looked at this and it _really is_ a can of worms (at least from the way I intended to solve it)

In anycase we should treat both separately - I think I agree with `external_sources` and `stub_paths` sounds like it can stay as it is.

**Comment by camilledejoye** [comment]
I guess I see the workspace folders as yet another feature.
Taking phpactor as an example I would have one project per per github repository, there is the `phpactor` project, `indexer-extension` project, `completion`project, etc.
But they are not meant to depends on each over, I can just switch from one to another while the client/server keep tracks of where I am.
So for each worskpace you have a different index but you only ever query the index matching your current workspace.

The current feature is more about having an external file or library that is shared by multiple projects so they are not included directly in any of them. Like [composer file autoload](https://getcomposer.org/doc/04-schema.md#files)
So for each workspace you have to include these files to the current index as if it was part of the current project.

So do we agree that to respond to the current issue we need to add the external sources (called `external_roots`) to the index and make sure it's watched ?
For simplicity I revert all the renaming and we keep the `stub_paths` as it is, as a way to add things to the index without the overhead of watching it.

**Comment by dantleech** [comment]
The referred to issue is already solved with the existing `stub_paths` functionality? So we don't need to do anything I guess.

`external_roots` is a new feature which extends the actively watched folders - in principle I'm not against that, but wonder if it's not premature or if there are better things to invest our time in (depending on the complexity of the task)

**Comment by camilledejoye** [comment]
OK, so I guess I can revert anything.
Maybe just keep the name resolving and update the doc to make it crystal clear that the index will be generated but not updated and a link to the index doc to show how to rebuild it in case of need.

**Comment by dantleech** [comment]
Thanks, I think this has helped to improve our understanding of the problem anyway :+1:

**Pull request closed without merging**